buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -1052,179 +1257,389 @@ private void transferToNextInLineOrRelease(Chunk chunk) {
+                    int remainingCapacity = curr.remainingCapacity();
+                    if (remainingCapacity > startingCapacity) {
+                        curr.readInitInto(buf, size, startingCapacity, maxCapacity);
+                        return true;
+                    }
+                    current = null;
+                    if (remainingCapacity >= size) {
+                        try {
+                            curr.readInitInto(buf, size, remainingCapacity, maxCapacity);
+                            return true;
+                        } finally {
+                            curr.releaseFromMagazine();
+                        }
+                    }
+                    if (remainingCapacity < RETIRE_CAPACITY) {
+                        curr.releaseFromMagazine();
+                    } else {
+                        transferToNextInLineOrRelease(curr);
+                    }
+                }
+            }
+
+            // If we reach here, `current` was null or the chunk was full.
+            // We need to find a new chunk to allocate from.
+            // Note: computeBufferCapacity is only needed if we allocate a new BumpChunk.
+            int startingCapacity = chunkController.computeBufferCapacity(size, maxCapacity, reallocate);
+            return allocateAndRefillCurrentChunk(size, maxCapacity, buf, startingCapacity, currentThread);
+        }
+
+        private boolean allocateAndRefillCurrentChunk(int size, int maxCapacity, AdaptiveByteBuf buf,
+                                                      int startingCapacity, Thread currentThread) {
+            Chunk curr = NEXT_IN_LINE.getAndSet(this, null);
+
+            if (curr != null) {
+                if (curr instanceof SizeClassedChunk) {
+                    if (((SizeClassedChunk) curr).tryAllocate(buf, size, maxCapacity)) {
+                        current = curr;
+                        return true;
+                    }
+                    // tryAllocate failed, chunk is full. Release it and fall through.
+                    curr.releaseFromMagazine();
+                } else {
+                    int remainingCapacity = curr.remainingCapacity();
+                    if (remainingCapacity > startingCapacity) {
+                        curr.readInitInto(buf, size, startingCapacity, maxCapacity);
+                        current = curr;
+                        return true;
+                    }
+                    if (remainingCapacity >= size) {
+                        try {
+                            curr.readInitInto(buf, size, remainingCapacity, maxCapacity);
+                            return true;
+                        } finally {
+                            curr.releaseFromMagazine();
+                        }
+                    } else {
+                        curr.releaseFromMagazine();
+                    }
```

Comment:
These 2 logic could be grouped under a single method namewhich have 2 impls in each chunk class type?